src/node_types/sequencer.rs

```diff
@@ -252,126 +251,113 @@
+                    DatabaseError::NotFoundError(format!("hashchain for ID {}: {}", id, e))
+                })?;
 
-        let mut tree = self.tree.lock().await;
-        let hashed_id = sha256_mod(id.as_bytes());
+                let mut tree = self.tree.lock().await;
+                let hashed_id = sha256_mod(id.as_bytes());
 
-        // todo: not all error cases make it okay to continue here, so we should filter by a Hashchain key not found error
-        if hashchain.is_ok() {
-            let node = match tree.find_leaf_by_label(&hashed_id) {
-                Some(node) => node,
-                None => {
+                let node = tree.find_leaf_by_label(&hashed_id).ok_or_else(|| {
                     // TODO: before merging, change error type
-                    return Err(GeneralError::DecodingError(format!(
+                    GeneralError::DecodingError(format!(
                         "node with label {} not found in the tree",
                         hashed_id
                     ))
-                    .into());
-                }
-            };
-            let new_index = match tree.find_node_index(&node) {
-                Some(index) => index,
-                None => {
-                    return Err(GeneralError::DecodingError(format!(
+                })?;
+
+                let previous_hash = current_chain.last().unwrap().hash;
+
+                let new_chain_entry = HashchainEntry::new(operation.clone(), previous_hash);
+                current_chain.push(new_chain_entry.clone());
+
+                let updated_node = Node::new_leaf(
+                    node.is_left_sibling(),
+                    hashed_id,
+                    new_chain_entry.hash,
+                    node.get_next(),
+                );
+
+                let index = tree.find_node_index(&node).ok_or_else(|| {
+                    GeneralError::DecodingError(format!(
                         "node with label {} not found in the tree, but has a hashchain entry",
                         hashed_id
                     ))
+                })?;
+
+                self.db
+                    .update_hashchain(operation, &current_chain)
+                    .map_err(|e| {
+                        PrismError::Database(DatabaseError::WriteError(format!(
+                            "hashchain for incoming operation {:?}: {:?}",
+                            operation, e
+                        )))
+                    })?;
+
+                // TODO: Possible optimization: cache the last update proof for each id for serving the proofs
+                tree.update_node(index, updated_node)
+                    .map(Proof::Update)
+                    .map_err(|e| e.into())
+            }
+            Operation::CreateAccount { id, value, source } => {
+                // validation of account source
+                match source {
+                    // TODO: use Signature, not String
```

Comment:
Please open a GH issue for this. Specifically, common::AccountSource::SignedBySequencer should use a zk_snark::Signature instead of a String. For this, the zk_snark::Signature type must be created.